### DISCUSSION TITLE
Possible compiler bug - no warning when calling extension method w/o await

### DISCUSSION THREAD

**Discussion by TahirAhmadov** [dotnet/csharplang#6176] [General] [answered]
```cs
#nullable enable
using System;
using System.Threading.Tasks;

public class C {
    ActionAsync? _action;
    public async Task M() {
        await Task.Delay(100);
        this._action?.InvokeAsync(); // no warning here (this caused a bug for me)
        if(this._action != null)
            this._action(); // however, this one does get a warning (as expected)
    }
}

public delegate Task ActionAsync();

public static partial class _Delegate_Extensions
{
	public static async Task InvokeAsync(this ActionAsync? This)
	{
		if (This != null)
		{
			foreach (ActionAsync d in This.GetInvocationList())
			{
				await d();
			}
		}
	}
}
```
VS does offer to add an await on that `this._action?.InvokeAsync();` line (but no warning).

PS. Never mind - I figured it out: I mistakenly added a `?` after `this._action` to make it null-propagating, which converted the return of `InvokeAsync` to `Task?`, hence no warning.
Hopefully I don't need this `InvokeAsync` with nullable `this` parameter anymore when we get "await if Task not null" feature.

**Comment by HaloFour** [comment, marked as answer]
Looks like it's the result of the null-propagation operator, removing that will emit the warning.  But an `await` there would throw if `_action` is `null`, so it's not exactly a good pattern in general, at least without `await?` or the like.

**Reply by TahirAhmadov** [reply]
I figured it out myself literally a minute after posting it. Yes we really need the `await?`.

**Reply by TahirAhmadov** [reply]
BTW, thanks for your response anyway.
